File: src/main/java/net/yakclient/opengl/render/GLRenderingData.java
package net.yakclient.opengl.render;

import net.yakclient.opengl.util.*;

import java.nio.DoubleBuffer;
import java.nio.FloatBuffer;

public class GLRenderingData {
    private final VerticeAggregation vertices;
    private final ColorAggregation colors;
    private final NormalAggregation normals;
    private final TexAggregation texs;

    private final Texture texture;

    private final int verticeCount;

    public GLRenderingData(VerticeAggregation vertices, ColorAggregation colors, NormalAggregation normals, TexAggregation texs, Texture texture, int verticeCount) {
        this.vertices = vertices;
        this.colors = colors;
        this.normals = normals;
        this.texs = texs;
        this.texture = texture;
        this.verticeCount = verticeCount;
    }

    public GLRenderingData(RenderingDataBuilder builder) {
        this(builder.vertices, builder.colors, builder.normals, builder.texs, builder.texture, builder.verticeCount);
    }

    public DoubleBuffer getVertices() {
        return this.vertices.asBuf(this.vertices.createBuf());
    }

    public DoubleBuffer getNormals() {
        return this.normals.asBuf(this.normals.createBuf());
    }

    public FloatBuffer getColors() {
        return this.colors.asBuf(this.colors.createBuf());
    }

    public FloatBuffer getTexs() {
        return this.texs.asBuf(this.texs.createBuf());
    }

    public Texture getTexture() {
        return texture;
    }

    public int getVerticeCount() {
        return verticeCount;
    }

    public static RenderingDataBuilder create() {
        return new RenderingDataBuilder();
    }

    public static RenderingDataBuilder create(Texture texture) {
        return new RenderingDataBuilder(texture);
    }

    public static class RenderingDataBuilder {
        private final VerticeAggregation vertices;
        private final ColorAggregation colors;
        private final NormalAggregation normals;
        private final TexAggregation texs;

        private final Texture texture;

        private int verticeCount = 0;


        public RenderingDataBuilder() {
            this.vertices = new VerticeAggregation();
            this.colors = new ColorAggregation();
            this.normals = new NormalAggregation();
            this.texs = new TexAggregation();
            this.texture = new VacantTexture();
        }

        public RenderingDataBuilder(Texture texture) {
            this.vertices = new VerticeAggregation();
            this.colors = new ColorAggregation();
            this.normals = new NormalAggregation();
            this.texs = new TexAggregation();
            this.texture = texture;
        }

        public RenderingDataBuilder addColor3f(float r, float g, float b) {
            this.colors.add(r, g, b);
            return this;
        }

        public RenderingDataBuilder addColor4f(float r, float g, float b, float alpha) {
            this.colors.add(r, b, b, alpha);
            return this;
        }

        public RenderingDataBuilder addColors(ColorAggregation aggregation) {
            this.colors.combine(aggregation);
            return this;
        }

        public RenderingDataBuilder addVertice2d(double x, double y) {
            this.verticeCount++;
            this.vertices.add(x, y);
            return this;
        }

        public RenderingDataBuilder addVertice3d(double x, double y, double z) {
            this.verticeCount++;
            this.vertices.add(x, y, z);
            return this;
        }

        public RenderingDataBuilder addVertices(VerticeAggregation aggregation) {
            this.verticeCount += aggregation.size();
            this.vertices.combine(aggregation);
            return this;
        }

        public RenderingDataBuilder addNormal(double x, double y, double z) {
            this.normals.add(x, y, z);
            return this;
        }

        public RenderingDataBuilder addNormals(NormalAggregation aggregation) {
            this.normals.combine(aggregation);
            return this;
        }

        public RenderingDataBuilder addTexVertice(float t, float s) {
            this.texs.add(t, s);
            return this;
        }

        public RenderingDataBuilder addTexs(TexAggregation aggregation) {
            this.texs.combine(aggregation);
            return this;
        }

        public GLRenderingData build() {
            return new GLRenderingData(this);
        }
    }
}

File: opengl2/src/main/kotlin/net/yakclient/graphics/opengl2/render/GLRenderingData.kt
package net.yakclient.graphics.opengl2.render

import net.yakclient.graphics.util.*
import java.nio.DoubleBuffer
import java.nio.FloatBuffer

/**
 * GLRendering data is considered actual data that needs to be
 * passed to opengl. It could for example be anything in a buffer
 * or a texture bound to opengl. It maintains data like vertice
 * count.
 *
 *
 * `GLRenderingContext` however doesnt contain what we consider
 * data but contains settings for opengl to process this data with.
 *
 * @author Durgan McBroom
 * @see VerticeRenderingContext
 *
 * @see Aggregation
 *
 * @see YakTexture
 */
public class GLRenderingData @JvmOverloads constructor(
    private val vertices: VerticeAggregation,
    private val colors: ColorAggregation = ColorAggregation(),
    private val normals: NormalAggregation = NormalAggregation(),
    private val texs: TexAggregation = TexAggregation(),
    public val texture: YakTexture = VacantTexture(),
    public val verticeCount: Int = vertices.size,
) {
    public val verticeBuf: DoubleBuffer by lazy { vertices.asBuffer()}
    public val colorBuf: FloatBuffer by lazy { colors.asBuffer() }
    public val normalBuf: DoubleBuffer by lazy { normals.asBuffer() }
    public val texBuf: FloatBuffer by lazy { texs.asBuffer() }

    public fun hasVertices(): Boolean = !vertices.isEmpty()
    public fun hasColors(): Boolean = !colors.isEmpty()
    public fun hasNormals(): Boolean = !normals.isEmpty()
    public fun hasTexs(): Boolean = !texs.isEmpty()
}